hspace.1
.TH HSPACE 1 2009-06-01 htools "Ganeti H-tools"
.SH NAME
hspace \- Cluster space analyzer for Ganeti

.SH SYNOPSIS
.B hspace
.B "[backend options...]"
.B "[algorithm options...]"
.B "[request options..."]
.BI "[ -p[" fields "] ]"
.B "[-v... | -q]"

.B hspace
.B --version

.TP
Backend options:
.BI " -m " cluster
|
.BI " -L[" path "]"
|
.BI " -t " data-file
|
.BI " --simulate " spec

.TP
Algorithm options:
.BI "[ --max-cpu " cpu-ratio " ]"
.BI "[ --min-disk " disk-ratio " ]"
.BI "[ -O " name... " ]"

.TP
Request options:
.BI "[--memory " mem "]"
.BI "[--disk " disk "]"
.BI "[--req-nodes " req-nodes "]"
.BI "[--vcpus " vcpus "]"
.BI "[--tiered-alloc " spec "]"


.SH DESCRIPTION
hspace computes how many additional instances can be fit on a cluster,
while maintaining N+1 status.

The program will try to place instances, all of the same size, on the
cluster, until the point where we don't have any N+1 possible
allocation. It uses the exact same allocation algorithm as the hail
iallocator plugin.

The output of the program is designed to interpreted as a shell
fragment (or parsed as a \fIkey=value\fR file). Options which extend
the output (e.g. \-p, \-v) will output the additional information on
stderr (such that the stdout is still parseable).

The following keys are available in the output of the script (all
prefixed with \fIHTS_\fR):
.TP
.I SPEC_MEM, SPEC_DSK, SPEC_CPU, SPEC_RQN
These represent the specifications of the instance model used for
allocation (the memory, disk, cpu, requested nodes).

.TP
.I CLUSTER_MEM, CLUSTER_DSK, CLUSTER_CPU, CLUSTER_NODES
These represent the total memory, disk, CPU count and total nodes in
the cluster.

.TP
.I INI_SCORE, FIN_SCORE
These are the initial (current) and final cluster score (see the hbal
man page for details about the scoring algorithm).

.TP
.I INI_INST_CNT, FIN_INST_CNT
The initial and final instance count.

.TP
.I INI_MEM_FREE, FIN_MEM_FREE
The initial and final total free memory in the cluster (but this
doesn't necessarily mean available for use).

.TP
.I INI_MEM_AVAIL, FIN_MEM_AVAIL
The initial and final total available memory for allocation in the
cluster. If allocating redundant instances, new instances could
increase the reserved memory so it doesn't necessarily mean the
entirety of this memory can be used for new instance allocations.

.TP
.I INI_MEM_RESVD, FIN_MEM_RESVD
The initial and final reserved memory (for redundancy/N+1 purposes).

.TP
.I INI_MEM_INST, FIN_MEM_INST
The initial and final memory used for instances (actual runtime used
RAM).

.TP
.I INI_MEM_OVERHEAD, FIN_MEM_OVERHEAD
The initial and final memory overhead \(em memory used for the node
itself and unacounted memory (e.g. due to hypervisor overhead).

.TP
.I INI_MEM_EFF, HTS_INI_MEM_EFF
The initial and final memory efficiency, represented as instance
memory divided by total memory.

.TP
.I INI_DSK_FREE, INI_DSK_AVAIL, INI_DSK_RESVD, INI_DSK_INST, INI_DSK_EFF
Initial disk stats, similar to the memory ones.

.TP
.I FIN_DSK_FREE, FIN_DSK_AVAIL, FIN_DSK_RESVD, FIN_DSK_INST, FIN_DSK_EFF
Final disk stats, similar to the memory ones.

.TP
.I INI_CPU_INST, FIN_CPU_INST
Initial and final number of virtual CPUs used by instances.

.TP
.I INI_CPU_EFF, FIN_CPU_EFF
The initial and final CPU efficiency, represented as the count of
virtual instance CPUs divided by the total physical CPU count.

.TP
.I INI_MNODE_MEM_AVAIL, FIN_MNODE_MEM_AVAIL
The initial and final maximum per\(hynode available memory. This is not
very useful as a metric but can give an impression of the status of
the nodes; as an example, this value restricts the maximum instance
size that can be still created on the cluster.

.TP
.I INI_MNODE_DSK_AVAIL, FIN_MNODE_DSK_AVAIL
Like the above but for disk.

.TP
.I TSPEC
If the tiered allocation mode has been enabled, this parameter holds
the pairs of specifications and counts of instances that can be
created in this mode. The value of the key is a space\(hyseparated list
of values; each value is of the form \fImemory,disk,vcpu=count\fR
where the memory, disk and vcpu are the values for the current spec,
and count is how many instances of this spec can be created. A
complete value for this variable could be: \fB4096,102400,2=225
2560,102400,2=20 512,102400,2=21\fR.

.TP
.I KM_USED_CPU, KM_USED_MEM, KM_USED_DSK
These represents the metrics of used resources at the start of the
computation (only for tiered allocation mode).

.TP
.I KM_POOL_CPU, KM_POOL_MEM, KM_POOL_DSK
These represents the total resources allocated during the tiered
allocation process. In effect, they represent how much is readily
available for allocation.

.TP
.I KM_UNAV_CPU, KM_UNAV_MEM, KM_UNAV_DSK
These represents the resources left over (either free as in
unallocable or allocable on their own) after the tiered allocation has
been completed. They represent better the actual unallocable
resources, because some other resource has been exhausted. For
example, the cluster might still have 100GiB disk free, but with no
memory left for instances, we cannot allocate another instance, so in
effect the disk space is unallocable. Note that the CPUs here
represent instance virtual CPUs, and in case the \fI--max-cpu\fR
option hasn't been specified this will be \-1.

.TP
.I ALLOC_USAGE
The current usage represented as initial number of instances divided
per final number of instances.

.TP
.I ALLOC_COUNT
The number of instances allocated (delta between FIN_INST_CNT and
INI_INST_CNT).

.TP
.I ALLOC_FAIL*_CNT
For the last attemp at allocations (which would have increased
FIN_INST_CNT with one, if it had succeeded), this is the count of the
failure reasons per failure type; currently defined are FAILMEM,
FAILDISK and FAILCPU which represent errors due to not enough memory,
disk and CPUs, and FAILN1 which represents a non N+1 compliant cluster
on which we can't allocate instances at all.

.TP
.I ALLOC_FAIL_REASON
The reason for most of the failures, being one of the above FAIL*
strings.

.TP
.I OK
A marker representing the successful end of the computation, and
having value "1". If this key is not present in the output it means
that the computation failed and any values present should not be
relied upon.

.PP

If the tiered allocation mode is enabled, then many of the INI_/FIN_
metrics will be also displayed with a TRL_ prefix, and denote the
cluster status at the end of the tiered allocation run.

.SH OPTIONS
The options that can be passed to the program are as follows:

.TP
.BI "--memory " mem
The memory size of the instances to be placed (defaults to 4GiB).

.TP
.BI "--disk " disk
The disk size of the instances to be placed (defaults to 100GiB).

.TP
.BI "--req-nodes " num-nodes
The number of nodes for the instances; the default of two means
mirrored instances, while passing one means plain type instances.

.TP
.BI "--vcpus " vcpus
The number of VCPUs of the instances to be placed (defaults to 1).

.TP
.BI "--max-cpu " cpu-ratio
The maximum virtual\(hyto\(hyphysical cpu ratio, as a floating point
number between zero and one. For example, specifying \fIcpu-ratio\fR
as \fB2.5\fR means that, for a 4\(hycpu machine, a maximum of 10
virtual cpus should be allowed to be in use for primary instances. A
value of one doesn't make sense though, as that means no disk space
can be used on it.

.TP
.BI "--min-disk " disk-ratio
The minimum amount of free disk space remaining, as a floating point
number. For example, specifying \fIdisk-ratio\fR as \fB0.25\fR means
that at least one quarter of disk space should be left free on nodes.

.TP
.B -p, --print-nodes
Prints the before and after node status, in a format designed to allow
the user to understand the node's most important parameters.

It is possible to customise the listed information by passing a
comma\(hyseparated list of field names to this option (the field list
is currently undocumented), or to extend the default field list by
prefixing the additional field list with a plus sign. By default, the
node list will contain the following information:
.RS
.TP
.B F
a character denoting the status of the node, with '\-' meaning an
offline node, '*' meaning N+1 failure and blank meaning a good node
.TP
.B Name
the node name
.TP
.B t_mem
the total node memory
.TP
.B n_mem
the memory used by the node itself
.TP
.B i_mem
the memory used by instances
.TP
.B x_mem
amount memory which seems to be in use but cannot be determined why or
by which instance; usually this means that the hypervisor has some
overhead or that there are other reporting errors
.TP
.B f_mem
the free node memory
.TP
.B r_mem
the reserved node memory, which is the amount of free memory needed
for N+1 compliance
.TP
.B t_dsk
total disk
.TP
.B f_dsk
free disk
.TP
.B pcpu
the number of physical cpus on the node
.TP
.B vcpu
the number of virtual cpus allocated to primary instances
.TP
.B pcnt
number of primary instances
.TP
.B pcnt
number of secondary instances
.TP
.B p_fmem
percent of free memory
.TP
.B p_fdsk
percent of free disk
.TP
.B r_cpu
ratio of virtual to physical cpus
.TP
.B lCpu
the dynamic CPU load (if the information is available)
.TP
.B lMem
the dynamic memory load (if the information is available)
.TP
.B lDsk
the dynamic disk load (if the information is available)
.TP
.B lNet
the dynamic net load (if the information is available)
.RE

.TP
.BI "-O " name
This option (which can be given multiple times) will mark nodes as
being \fIoffline\fR, and instances won't be placed on these nodes.

Note that hspace will also mark as offline any nodes which are
reported by RAPI as such, or that have "?" in file\(hybased input in any
numeric fields.
.RE

.TP
.BI "-t" datafile ", --text-data=" datafile
The name of the file holding node and instance information (if not
collecting via RAPI or LUXI). This or one of the other backends must
be selected.

.TP
.BI "-S" filename ", --save-cluster=" filename
If given, the state of the cluster at the end of the allocation is
saved to a file named \fIfilename.alloc\fR, and if tiered allocation
is enabled, the state after tiered allocation will be saved to
\fIfilename.tiered\fR. This allows re-feeding the cluster state to
either hspace itself (with different parameters) or for example hbal.

.TP
.BI "-m" cluster
Collect data directly from the
.I cluster
given as an argument via RAPI. If the argument doesn't contain a colon
(:), then it is converted into a fully\(hybuilt URL via prepending
https:// and appending the default RAPI port, otherwise it's
considered a fully\(hyspecified URL and is used as\(hyis.

.TP
.BI "-L[" path "]"
Collect data directly from the master daemon, which is to be contacted
via the luxi (an internal Ganeti protocol). An optional \fIpath\fR
argument is interpreted as the path to the unix socket on which the
master daemon listens; otherwise, the default path used by ganeti when
installed with \fI--localstatedir=/var\fR is used.

.TP
.BI "--simulate " description
Instead of using actual data, build an empty cluster given a node
description. The \fIdescription\fR parameter must be a
comma\(hyseparated list of four elements, describing in order:

.RS

.RS
.TP
the number of nodes in the cluster

.TP
the disk size of the nodes, in mebibytes

.TP
the memory size of the nodes, in mebibytes

.TP
the cpu core count for the nodes

.RE

An example description would be \fB20,102400,16384,4\fR describing a
20\(hynode cluster where each node has 100GiB of disk space, 16GiB of
memory and 4 CPU cores. Note that all nodes must have the same specs
currently.

.RE

.TP
.BI "--tiered-alloc " spec
Beside the standard, fixed\(hysize allocation, also do a tiered
allocation scheme where the algorithm starts from the given
specification and allocates until there is no more space; then it
decreases the specification and tries the allocation again. The
decrease is done on the matric that last failed during allocation. The
specification given is similar to the \fI--simulate\fR option and it
holds:

.RS

.RS

.TP
the disk size of the instance

.TP
the memory size of the instance

.TP
the vcpu count for the insance

.RE

An example description would be \fB10240,8192,2\fR describing an
initial starting specification of 10GiB of disk space, 4GiB of memory
and 2 VCPUs.

Also note that the normal allocation and the tiered allocation are
independent, and both start from the initial cluster state; as such,
the instance count for these two modes are not related one to another.

.RE

.TP
.B -v, --verbose
Increase the output verbosity. Each usage of this option will increase
the verbosity (currently more than 2 doesn't make sense) from the
default of one. At verbosity 2 the location of the new instances is
shown in the standard error.

.TP
.B -q, --quiet
Decrease the output verbosity. Each usage of this option will decrease
the verbosity (less than zero doesn't make sense) from the default of
one.

.TP
.B -V, --version
Just show the program version and exit.

.SH EXIT STATUS

The exist status of the command will be zero, unless for some reason
the algorithm fatally failed (e.g. wrong node or instance data).

.SH BUGS

The algorithm is highly dependent on the number of nodes; its runtime
grows exponentially with this number, and as such is impractical for
really big clusters.

The algorithm doesn't rebalance the cluster or try to get the optimal
fit; it just allocates in the best place for the current step, without
taking into consideration the impact on future placements.

.SH ENVIRONMENT

If the variables \fBHTOOLS_NODES\fR and \fBHTOOLS_INSTANCES\fR are
present in the environment, they will override the default names for
the nodes and instances files. These will have of course no effect
when the RAPI or Luxi backends are used.

.SH SEE ALSO
.BR hbal "(1), " hscan "(1), " ganeti "(7), " gnt-instance "(8), "
.BR gnt-node "(8)"

.SH "COPYRIGHT"
.PP
Copyright (C) 2009 Google Inc. Permission is granted to copy,
distribute and/or modify under the terms of the GNU General Public
License as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.
.PP
On Debian systems, the complete text of the GNU General Public License
can be found in /usr/share/common-licenses/GPL.


NEWS
Ganeti-htools release notes
===========================


Version 0.2.7 (Unreleased)
--------------------------

Bug fixes:

- fixed the error message for hail multi-evacuation mode

New features:

- add a new option ``-S`` to hbal and hspace that saves the cluster
  state at the end of the processing in the text format used by the
  ``-t`` option, for later re-processing

Version 0.2.6 (Mon, 26 Jul 2010)
--------------------------------

Exactly three months since the last release. Many internal changes, plus
a couple of important changes in the balancing algorithm.

First, the balancing may now introduce N+1 errors, if this solves other,
more critical problems. For the moment, this means that moving instances
away from offline nodes is allowed even if it creates N+1 errors, and
that means evacuation can be done in more cases.

Second, the scoring for N+1 has changed. In previous versions, it simply
counted the number of failing N+1 nodes, which means moving an instance
away from a N+1 failed node (but without the node 'clearing' the N+1
status) was not reflected in the cluster score. As such, the balancing
algorithm managed to clear N+1 errors only sometimes, since usually it
takes more than one move for this, and the first prerequisite move was
not 'rewarded' appropriately and thus it was not selected. Now, it is
possible to fix many more error cases than before: on a simulated 40
node cluster full with instances (symmetrically allocated on all nodes),
around five nodes can be evacuated before N+1 errors can be solved,
whereas 0.2.5 could evacuate at best one node.

There were some other internal changes to the scoring algorithm, such
that now the metrics have associated weights, and they are not all of
the same importance anymore. As of now, the only change is that offline
instances have a higher weight, which should favour proper node
evacuations.

Among the other changes:

- fixed the hspace KM_POOL_* metrics, which were returned as the final
  state and not as the delta between the initial and final states
- fixed hspace handling of N+1 failing clusters: before, it used to
  generate a 'fake' response, and the structure of this response was not
  always in sync with the real responses, leading to missing items;
  currently it proceeds correctly through the code (skipping the
  computation), and uses the same display mechanisms as the normal case
- fixed hscan exit code for RAPI failures: previously it finished with
  success even if all the clusters failed, which was creating issues
  with the live-test script; now it exits with exit code 2 for RAPI
  failures (unfortunately this is still not optimal as LUXI failures
  will use exit code 1, the same as the command line)
- changed the limit values for CPU/disk, which previously were used
  optionally, whereas now they are always used; the default cpu ratio
  limit is now 64 VCPUs per PCPU
- changed the internal handling of the short name vs. original
  (Ganeti-provided) name; now internally we always use the full name,
  and only in display routines we show the shortened (called 'alias')
  name; as a result, the -O and --excluded-instances options now accept
  both the full name and the shortened name
- changed internal handling of JSON conversions and errors, such that
  now we show a better context for failure messages, which should help
  with diagnosing the malformed message
- changed the names for a few node fields, and added some more nodes;
  this is most likely to help with debugging, and not with regular
  operation though
- changed the node fields option to allow the '+' prefix to mean 'extend
  the default fields list' rather than start from fresh (similar to
  Ganeti's implementation)
- a few internal changes related to the LUXI protocol implementation,
  which should make it more safe against potential bugs, one
  optiomization that should help with large messages, and some patches
  in preparation for potential expansion of the LUXI backend functionality

And finally, many improvements on unittests and the live-test
script. Test coverage is much enhanced, and the test infrastructure has
better error reporting; this should lead down-the-road to better code
and fewer bugs…


Version 0.2.5 (Mon, 26 Apr 2010)
--------------------------------

Some internal cleanup plus a few user-visible changes:

- new option for marking instances as 'do-not-move' during rebalancing
- allow ``hscan`` to scan the local cluster via Luxi
- add more metrics to ``hspace`` which show the delta between original
  state and final state better (only valid for tiered allocation)


Version 0.2.4 (Mon, 22 Feb 2010)
--------------------------------

Two improvements for node evacuation:

- hbal takes a new parameter ``--evac-mode`` that restricts the
  instances to be moved to the ones on offline/drained nodes, which
  should reduce the work done
- hail supports the new ``multi-evacuate`` mode of the IAllocator
  protocol, that will be released in a minor release on the Ganeti 2.1
  branch


Version 0.2.3 (Thu,  4 Feb 2010)
--------------------------------

A small release:

- Fixes selection of secondary node: previously, if the cluster had
  many N+1 failures, a N+1 failed node could be selected as secondary
  even if it did not have enough memory to allow the instance to be
  migrated/failed over to it; this is bad for automated tools, since
  we can get the cluster in an unhealthy state
- Switch the text backend to a single input file, that is generated
  now by hscan and shouldn't be generated manually via
  gnt-node/instance list anymore; this allows richer information to be
  kept in the file, and simplifies a little the internals of the text
  backend


Version 0.2.2 (Tue, 29 Dec 2009)
--------------------------------

Small release, 0.2.1 was broken and thus this was released earlier:

- Release 0.2.1 broke the LUXI backend due to a typo, fixed
- Added a live-test script that should catch errors like the above one

hspace.hs
{-| Cluster space sizing

-}

{-

Copyright (C) 2009, 2010 Google Inc.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301, USA.

-}

module Main (main) where

import Data.Char (toUpper, isAlphaNum)
import Data.List
import Data.Function
import Data.Maybe (isJust, fromJust)
import Data.Ord (comparing)
import Monad
import System (exitWith, ExitCode(..))
import System.FilePath
import System.IO
import qualified System

import Text.Printf (printf, hPrintf)

import qualified Ganeti.HTools.Container as Container
import qualified Ganeti.HTools.Cluster as Cluster
import qualified Ganeti.HTools.Node as Node
import qualified Ganeti.HTools.Instance as Instance

import Ganeti.HTools.Utils
import Ganeti.HTools.Types
import Ganeti.HTools.CLI
import Ganeti.HTools.ExtLoader
import Ganeti.HTools.Text (serializeCluster)

-- | Options list and functions
options :: [OptType]
options =
    [ oPrintNodes
    , oDataFile
    , oNodeSim
    , oRapiMaster
    , oLuxiSocket
    , oVerbose
    , oQuiet
    , oOfflineNode
    , oIMem
    , oIDisk
    , oIVcpus
    , oINodes
    , oMaxCpu
    , oMinDisk
    , oTieredSpec
    , oSaveCluster
    , oShowVer
    , oShowHelp
    ]

-- | The allocation phase we're in (initial, after tiered allocs, or
-- after regular allocation).
data Phase = PInitial
           | PFinal
           | PTiered

statsData :: [(String, Cluster.CStats -> String)]
statsData = [ ("SCORE", printf "%.8f" . Cluster.csScore)
            , ("INST_CNT", printf "%d" . Cluster.csNinst)
            , ("MEM_FREE", printf "%d" . Cluster.csFmem)
            , ("MEM_AVAIL", printf "%d" . Cluster.csAmem)
            , ("MEM_RESVD",
               \cs -> printf "%d" (Cluster.csFmem cs - Cluster.csAmem cs))
            , ("MEM_INST", printf "%d" . Cluster.csImem)
            , ("MEM_OVERHEAD",
               \cs -> printf "%d" (Cluster.csXmem cs + Cluster.csNmem cs))
            , ("MEM_EFF",
               \cs -> printf "%.8f" (fromIntegral (Cluster.csImem cs) /
                                     Cluster.csTmem cs))
            , ("DSK_FREE", printf "%d" . Cluster.csFdsk)
            , ("DSK_AVAIL", printf "%d". Cluster.csAdsk)
            , ("DSK_RESVD",
               \cs -> printf "%d" (Cluster.csFdsk cs - Cluster.csAdsk cs))
            , ("DSK_INST", printf "%d" . Cluster.csIdsk)
            , ("DSK_EFF",
               \cs -> printf "%.8f" (fromIntegral (Cluster.csIdsk cs) /
                                    Cluster.csTdsk cs))
            , ("CPU_INST", printf "%d" . Cluster.csIcpu)
            , ("CPU_EFF",
               \cs -> printf "%.8f" (fromIntegral (Cluster.csIcpu cs) /
                                     Cluster.csTcpu cs))
            , ("MNODE_MEM_AVAIL", printf "%d" . Cluster.csMmem)
            , ("MNODE_DSK_AVAIL", printf "%d" . Cluster.csMdsk)
            ]

specData :: [(String, RSpec -> String)]
specData = [ ("MEM", printf "%d" . rspecMem)
           , ("DSK", printf "%d" . rspecDsk)
           , ("CPU", printf "%d" . rspecCpu)
           ]

clusterData :: [(String, Cluster.CStats -> String)]
clusterData = [ ("MEM", printf "%.0f" . Cluster.csTmem)
              , ("DSK", printf "%.0f" . Cluster.csTdsk)
              , ("CPU", printf "%.0f" . Cluster.csTcpu)
              , ("VCPU", printf "%d" . Cluster.csVcpu)
              ]

-- | Function to print stats for a given phase
printStats :: Phase -> Cluster.CStats -> [(String, String)]
printStats ph cs =
  map (\(s, fn) -> (printf "%s_%s" kind s, fn cs)) statsData
  where kind = case ph of
                 PInitial -> "INI"
                 PFinal -> "FIN"
                 PTiered -> "TRL"

-- | Print final stats and related metrics
printResults :: Node.List -> Int -> Int -> [(FailMode, Int)] -> IO ()
printResults fin_nl num_instances allocs sreason = do
  let fin_stats = Cluster.totalResources fin_nl
      fin_instances = num_instances + allocs

  when (num_instances + allocs /= Cluster.csNinst fin_stats) $
       do
         hPrintf stderr "ERROR: internal inconsistency, allocated (%d)\
                        \ != counted (%d)\n" (num_instances + allocs)
                                 (Cluster.csNinst fin_stats) :: IO ()
         exitWith $ ExitFailure 1

  printKeys $ printStats PFinal fin_stats
  printKeys [ ("ALLOC_USAGE", printf "%.8f"
                                ((fromIntegral num_instances::Double) /
                                 fromIntegral fin_instances))
            , ("ALLOC_INSTANCES", printf "%d" allocs)
            , ("ALLOC_FAIL_REASON", map toUpper . show . fst $ head sreason)
            ]
  printKeys $ map (\(x, y) -> (printf "ALLOC_%s_CNT" (show x),
                               printf "%d" y)) sreason
  -- this should be the final entry
  printKeys [("OK", "1")]

formatRSpec :: String -> RSpec -> [(String, String)]
formatRSpec s r =
    [ ("KM_" ++ s ++ "_CPU", show $ rspecCpu r)
    , ("KM_" ++ s ++ "_MEM", show $ rspecMem r)
    , ("KM_" ++ s ++ "_DSK", show $ rspecDsk r)
    ]

printAllocationStats :: Node.List -> Node.List -> IO ()
printAllocationStats ini_nl fin_nl = do
  let ini_stats = Cluster.totalResources ini_nl
      fin_stats = Cluster.totalResources fin_nl
      (rini, ralo, runa) = Cluster.computeAllocationDelta ini_stats fin_stats
  printKeys $ formatRSpec "USED" rini
  printKeys $ formatRSpec "POOL" ralo
  printKeys $ formatRSpec "UNAV" runa

-- | Ensure a value is quoted if needed
ensureQuoted :: String -> String
ensureQuoted v = if not (all (\c -> (isAlphaNum c || c == '.')) v)
                 then '\'':v ++ "'"
                 else v

-- | Format a list of key\/values as a shell fragment
printKeys :: [(String, String)] -> IO ()
printKeys = mapM_ (\(k, v) ->
                   printf "HTS_%s=%s\n" (map toUpper k) (ensureQuoted v))

printInstance :: Node.List -> Instance.Instance -> [String]
printInstance nl i = [ Instance.name i
                     , Container.nameOf nl $ Instance.pNode i
                     , let sdx = Instance.sNode i
                       in if sdx == Node.noSecondary then ""
                          else Container.nameOf nl sdx
                     , show (Instance.mem i)
                     , show (Instance.dsk i)
                     , show (Instance.vcpus i)
                     ]

-- | Main function.
main :: IO ()
main = do
  cmd_args <- System.getArgs
  (opts, args) <- parseOpts cmd_args "hspace" options

  unless (null args) $ do
         hPutStrLn stderr "Error: this program doesn't take any arguments."
         exitWith $ ExitFailure 1

  let verbose = optVerbose opts
      ispec = optISpec opts
      shownodes = optShowNodes opts

  (fixed_nl, il, _) <- loadExternalData opts

  printKeys $ map (\(a, fn) -> ("SPEC_" ++ a, fn ispec)) specData
  printKeys [ ("SPEC_RQN", printf "%d" (optINodes opts)) ]

  let num_instances = length $ Container.elems il

  let offline_names = optOffline opts
      all_nodes = Container.elems fixed_nl
      all_names = map Node.name all_nodes
      offline_wrong = filter (`notElem` all_names) offline_names
      offline_indices = map Node.idx $
                        filter (\n ->
                                 Node.name n `elem` offline_names ||
                                 Node.alias n `elem` offline_names)
                               all_nodes
      req_nodes = optINodes opts
      m_cpu = optMcpu opts
      m_dsk = optMdsk opts

  when (length offline_wrong > 0) $ do
         hPrintf stderr "Error: Wrong node name(s) set as offline: %s\n"
                     (commaJoin offline_wrong) :: IO ()
         exitWith $ ExitFailure 1

  when (req_nodes /= 1 && req_nodes /= 2) $ do
         hPrintf stderr "Error: Invalid required nodes (%d)\n"
                                            req_nodes :: IO ()
         exitWith $ ExitFailure 1

  let nm = Container.map (\n -> if Node.idx n `elem` offline_indices
                                then Node.setOffline n True
                                else n) fixed_nl
      nl = Container.map (flip Node.setMdsk m_dsk . flip Node.setMcpu m_cpu)
           nm
      csf = commonSuffix fixed_nl il

  when (length csf > 0 && verbose > 1) $
       hPrintf stderr "Note: Stripping common suffix of '%s' from names\n" csf

  when (isJust shownodes) $
       do
         hPutStrLn stderr "Initial cluster status:"
         hPutStrLn stderr $ Cluster.printNodes nl (fromJust shownodes)

  let ini_cv = Cluster.compCV nl
      ini_stats = Cluster.totalResources nl

  when (verbose > 2) $
         hPrintf stderr "Initial coefficients: overall %.8f, %s\n"
                 ini_cv (Cluster.printStats nl)

  printKeys $ map (\(a, fn) -> ("CLUSTER_" ++ a, fn ini_stats)) clusterData
  printKeys [("CLUSTER_NODES", printf "%d" (length all_nodes))]
  printKeys $ printStats PInitial ini_stats

  let bad_nodes = fst $ Cluster.computeBadItems nl il
      stop_allocation = length bad_nodes > 0
      result_noalloc = ([(FailN1, 1)]::FailStats, nl, il, [])

  -- utility functions
  let iofspec spx = Instance.create "new" (rspecMem spx) (rspecDsk spx)
                    (rspecCpu spx) "running" [] (-1) (-1)
      exitifbad val = (case val of
                         Bad s -> do
                           hPrintf stderr "Failure: %s\n" s :: IO ()
                           exitWith $ ExitFailure 1
                         Ok x -> return x)


  let reqinst = iofspec ispec

  -- Run the tiered allocation, if enabled

  (case optTieredSpec opts of
     Nothing -> return ()
     Just tspec -> do
       (_, trl_nl, trl_il, trl_ixes) <-
           if stop_allocation
           then return result_noalloc
           else exitifbad (Cluster.tieredAlloc nl il (iofspec tspec)
                                  req_nodes [])
       let fin_trl_ixes = reverse trl_ixes
           ix_byspec = groupBy ((==) `on` Instance.specOf) fin_trl_ixes
           spec_map = map (\ixs -> (Instance.specOf $ head ixs, length ixs))
                      ix_byspec::[(RSpec, Int)]
           spec_map' = map (\(spec, cnt) ->
                                printf "%d,%d,%d=%d" (rspecMem spec)
                                       (rspecDsk spec) (rspecCpu spec) cnt)
                       spec_map::[String]

       when (verbose > 1) $ do
         hPutStrLn stderr "Tiered allocation map"
         hPutStr stderr . unlines . map ((:) ' ' .  intercalate " ") $
                 formatTable (map (printInstance trl_nl) fin_trl_ixes)
                                 [False, False, False, True, True, True]

       when (isJust shownodes) $ do
         hPutStrLn stderr ""
         hPutStrLn stderr "Tiered allocation status:"
         hPutStrLn stderr $ Cluster.printNodes trl_nl (fromJust shownodes)

       when (isJust $ optSaveCluster opts) $
            do
              let out_path = (fromJust $ optSaveCluster opts) <.> "tiered"
                  adata = serializeCluster trl_nl trl_il
              writeFile out_path adata
              hPrintf stderr "The cluster state after tiered allocation\
                             \ has been written to file '%s'\n"
                             out_path
       printKeys $ printStats PTiered (Cluster.totalResources trl_nl)
       printKeys [("TSPEC", intercalate " " spec_map')]
       printAllocationStats nl trl_nl)

  -- Run the standard (avg-mode) allocation

  (ereason, fin_nl, fin_il, ixes) <-
      if stop_allocation
      then return result_noalloc
      else exitifbad (Cluster.iterateAlloc nl il reqinst req_nodes [])

  let allocs = length ixes
      fin_ixes = reverse ixes
      sreason = reverse $ sortBy (comparing snd) ereason

  when (verbose > 1) $ do
         hPutStrLn stderr "Instance map"
         hPutStr stderr . unlines . map ((:) ' ' .  intercalate " ") $
                 formatTable (map (printInstance fin_nl) fin_ixes)
                                 [False, False, False, True, True, True]
  when (isJust shownodes) $
       do
         hPutStrLn stderr ""
         hPutStrLn stderr "Final cluster status:"
         hPutStrLn stderr $ Cluster.printNodes fin_nl (fromJust shownodes)

  when (isJust $ optSaveCluster opts) $
       do
         let out_path = (fromJust $ optSaveCluster opts) <.> "alloc"
             adata = serializeCluster fin_nl fin_il
         writeFile out_path adata
         hPrintf stderr "The cluster state after standard allocation\
                        \ has been written to file '%s'\n"
                 out_path

  printResults fin_nl num_instances allocs sreason
